feat!: convert to typescript

vogdb opened the pull request:
The PR is not ready for a merge. I've pushed it to ask some questions. The questions are posted as comments at the relevant source lines.

vogdb committed:
feat: convert to typescript

vogdb commented on package.json:
```diff
@@ -52,10 +71,9 @@
   "dependencies": {
+    "@libp2p/logger": "^1.0.2",
     "abortable-iterator": "^3.0.0",
-    "class-is": "^1.1.0",
```
I assumed that `class-is` is not required if the project is migrated to Typescript. Am I correct?

wemeetagain replied:
Yeah :+1:

vogdb commented on package.json:
```diff
@@ -40,6 +40,25 @@
+  ],
+  "exports": {
+    ".": {
+      "import": "./dist/src/index.js"
```
I followed other libp2p projects, and allowed exporting as ESM modules only.

wemeetagain replied:
sounds good to me

vogdb commented on src/socket-to-conn.ts:
```diff
@@ -49,11 +63,11 @@ module.exports = (socket, options = {}) => {
 
     // If the remote address was passed, use it - it may have the peer ID encapsulated
-    remoteAddr: options.remoteAddr,
+    remoteAddr: options.remoteAddr ?? toMultiaddr(socket.remoteAddress ?? '', socket.remotePort ?? ''),
```
I had to add this conversion because `remoteAddr` is required in `MultiaddrConnection` interface.

wemeetagain replied:
Maybe we should just throw if `options.remoteAddr` or `socket.remoteAddress` or `socket.remotePort` is nullish. Having a blank multiaddr or ill-formed multiaddr is not useful here.

vogdb replied:
I made `remoteAddr` option mandatory because all usages of `ToConnectionOptions` specify it. Please verify.

vogdb commented on src/listener.ts:
```diff
@@ -0,0 +1,145 @@
+    })
+
+    const maConn = toMultiaddrConnection(channel, {
+      remoteAddr: ipPortToMultiaddr(req.socket.remoteAddress, req.socket.remotePort)
```
Should we throw when `req.socket.remoteAddress` is undefined? It can be, according to the docs https://nodejs.org/api/net.html#class-netsocket

wemeetagain replied:
I'm not sure if it's necessary or if its just used for diagnostics, @vasco-santos any thoughts?

vogdb replied:
I made `remoteAddr` option mandatory because all usages of `ToConnectionOptions` specify it. Please verify.

vogdb commented on src/index.ts:
```diff
@@ -1,66 +1,54 @@
+const log = logger('libp2p:webrtcdirect')
+
+export interface WebRTCDirectListenerOptions extends ListenerOptions{
+  channelOptions?: Object
```
I followed the same logic as in [js-libp2p-webrtc-star/webrtc-star-transport](https://github.com/libp2p/js-libp2p-webrtc-star/blob/master/packages/webrtc-star-transport/src/index.ts#L65). However I didn't describe the type of `channelOptions`. As far as this code is concerned, `Object` type looked fine but let me know if you want to describe it thoroughly.

wemeetagain replied:
It would be better to describe the shape entirely, not use `Object` type.

vogdb replied:
Please look. The exact shape is described here https://www.npmjs.com/package/libp2p-webrtc-peer#api. See constructor options:
```
{
  initiator: false,
  channelConfig: {},
  channelName: '<random string>',
  config: { iceServers: [{ urls: 'stun:stun.l.google.com:19302' }, { urls: 'stun:global.stun.twilio.com:3478?transport=udp' }] },
  offerOptions: {},
  answerOptions: {},
  sdpTransform: function (sdp) { return sdp },
  stream: false,
  streams: [],
  trickle: true,
  allowHalfTrickle: false,
  wrtc: {}, // RTCPeerConnection/RTCSessionDescription/RTCIceCandidate
  objectMode: false
}
```
I described all the keys of this object. However for `config` key I used a plain object. Do you want me to describe it more rigid with `iceServers` and `urls`?

vogdb commented on src/index.ts:
```diff
@@ -1,66 +1,54 @@
-   * @returns {Promise<SimplePeer>} Resolves a SimplePeer Webrtc channel
-   */
-  _connect (ma, options = {}) {
+  async _connect (ma: Multiaddr, options:any ={}) {
```
I forgot to migrate `options` here. My bad.

vogdb replied:
done

vogdb committed:
introduce WebRTCDirectDialOptions

vogdb commented:
@vasco-santos @wemeetagain , please look when you have a minute.

vogdb committed:
mandatory options

- make `ToConnectionOptions` mandatory in `toMultiaddrConnection`
- make `remoteAddr` mandatory in `ToConnectionOptions`

vogdb committed:
Rename `channelOptions` to `peerOptions`. Describe it thoroughly.

achingbrain committed:
feat: convert to typescript

- Update code to use interfaces from `@libp2p/interfaces`
- Update module name to use the `@libp2p/` scope
- Use the same `@libp2p/webrtc-peer` module as `@libp2p/webrtc-star`
- Update licenses, CI and project config to match other `@libp2p/` modules

BREAKING CHANGE: this module now only exports named exports and is ESM only

achingbrain committed:
chore: catch stream errors

achingbrain committed:
chore: remove main

achingbrain committed:
chore: update release scripts

achingbrain committed:
chore: catch more errors when we close streams we are not interested in

achingbrain committed:
fix: remove catches from tests

achingbrain committed:
chore: implement skipped tests

achingbrain committed:
chore: update published files

achingbrain committed:
chore: slow ci is slow

achingbrain committed:
chore: remove extra timeouts

achingbrain committed:
chore: ci

achingbrain committed:
chore: ci

achingbrain committed:
chore: slow ci is not that slow, also add debug for windows

achingbrain committed:
chore: only debug during test run

achingbrain committed:
chore: just install

achingbrain committed:
chore: run all tests

achingbrain committed:
chore: ci y u no ci

achingbrain committed:
chore: update deps

achingbrain committed:
chore: ci

achingbrain committed:
chore: ci

achingbrain committed:
chore: remove debug

achingbrain commented:
I've updated this PR to use the new `@libp2p/*` typescript modules, including the `@libp2p/webrtc-peer` module that is used by `@libp2p/webrtc-star` - this is the bits of `simple-peer` that we use ported to typescript - e.g. it's just a one-to-one RTC data channel without support for audio or video since we don't use that.

It also has auto-release now, should be all set up to go.

achingbrain committed:
chore: ci

achingbrain committed:
chore: ci

achingbrain commented:
GH actions is being really flaky today, only seems to run on about 1/2 of the commits.

mpetrunic reviewed (commented):
lgtm
